feat(windows): method set context correctly truncates input buffer

rc-swag opened the pull request:
Fixes #6079 
AppContext Set now truncates an input buffer correctly if it is longer in length than the internal storage available for the context,
The same implementation is used as found in the common core found in kmx_context.cpp.

@keymanapp-test-bot skip

mcdurdin commented on windows/src/engine/keyman32/appint/appint.cpp:
```diff
@@ -135,16 +137,47 @@ void AppContext::CopyFrom(AppContext *source)   // I3575
+  pos = (int)(intptr_t)(q - CurContext);
+  CurContext[MAXCONTEXT - 1] = 0;
+
+  //////
+  //const WCHAR *p;
+  //WCHAR *q;
+  //for (p = buf, q = CurContext; *p && (INT_PTR)(q - CurContext) < MAXCONTEXT - 1; p++, q++) {
+  //  *q = *p;
+  //  if (*p >= 0xD800 && *p <= 0xDBFF) {
+  //    *(++q) = *(++p);
+  //  }
+  //}
+  //*q                         = 0;
+  //pos                        = (int)(INT_PTR)(q - CurContext);  // I1129 - Irregular behaviour with context rules
+  //CurContext[MAXCONTEXT - 1] = 0;
```
Please remove commented code

rc-swag replied:
Oh no that is a worry I know I removed that but must not off hit Ctrl+S before pushing it :(.

mcdurdin commented on windows/src/engine/keyman32/appint/appint.cpp:
```diff
@@ -41,10 +41,12 @@ AppContext::AppContext()
 {
-	if(pos == MAXCONTEXT - 1)
-	{
+	if(pos == MAXCONTEXT - 1){
```
```suggestion
	if(pos == MAXCONTEXT - 1) {
```

mcdurdin reviewed (changes requested):
LGTM except for the cleanup. Appreciate that this change should not be needed for long.

darcywong00 commented on windows/src/engine/keyman32/tests/keyman-engine-tests/appinttests.cpp:
```diff
@@ -6,7 +6,7 @@ TEST(AppContext, Get_Max_buff) {
   AppContext testContext;
   WCHAR *testString = L"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz0123456789";
-  WCHAR *expectedString = L"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz0"; // MAXCONTEXT taking into account null character termination
+  WCHAR *expectedString = L"JKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz0123456789"; // MAXCONTEXT taking into account null character termination
```
Why is the expected string truncating the beginning of the buffer? I thought it would truncate the end...

rc-swag replied:
The AppContext Class has an array of 16bit Unicode characters (CurContext) that stores the current context it is limited to 64. The Character at index 0 into the array is the furthest from the caret.  So when truncating we are interested in the text closet to the caret, therefore we truncate the begging of the buffer.

rc-swag commented:
> LGTM except for the cleanup. Appreciate that this change should not be needed for long.

I have done a clean up merging is still blocked I think I have addressed review comments.

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 15.0.176-alpha](https://keyman.com/downloads/releases/alpha/15.0.176)
